<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 1 Short reads mapping | Standardized pipelines</title>
  <meta name="description" content="This book contains the standardized pipelines used in the HaploTeam." />
  <meta name="generator" content="bookdown 0.32 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 1 Short reads mapping | Standardized pipelines" />
  <meta property="og:type" content="book" />
  
  <meta property="og:description" content="This book contains the standardized pipelines used in the HaploTeam." />
  <meta name="github-repo" content="rstudio/bookdown-demo" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 1 Short reads mapping | Standardized pipelines" />
  
  <meta name="twitter:description" content="This book contains the standardized pipelines used in the HaploTeam." />
  

<meta name="author" content="Victor Loegler" />


<meta name="date" content="2023-01-25" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="index.html"/>
<link rel="next" href="snps-and-indels-calling.html"/>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.1.0/anchor-sections.css" rel="stylesheet" />
<link href="libs/anchor-sections-1.1.0/anchor-sections-hash.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.1.0/anchor-sections.js"></script>
<script src="libs/clipboard-1.7.1/clipboard.min.js"></script>
<link href="libs/primer-tooltips-1.4.0/build.css" rel="stylesheet" />
<link href="libs/klippy-0.0.0.9500/css/klippy.min.css" rel="stylesheet" />
<script src="libs/klippy-0.0.0.9500/js/klippy.min.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>


<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">HaploTeam's Standardized Pipelines</a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Introduction</a></li>
<li class="chapter" data-level="1" data-path="short-reads-mapping.html"><a href="short-reads-mapping.html"><i class="fa fa-check"></i><b>1</b> Short reads mapping</a></li>
<li class="chapter" data-level="2" data-path="snps-and-indels-calling.html"><a href="snps-and-indels-calling.html"><i class="fa fa-check"></i><b>2</b> SNPs and INDELs calling</a></li>
<li class="chapter" data-level="3" data-path="snp-and-indel-filtering.html"><a href="snp-and-indel-filtering.html"><i class="fa fa-check"></i><b>3</b> SNP and INDEL filtering</a>
<ul>
<li class="chapter" data-level="3.1" data-path="snp-and-indel-filtering.html"><a href="snp-and-indel-filtering.html#quality-filtering"><i class="fa fa-check"></i><b>3.1</b> Quality filtering</a></li>
<li class="chapter" data-level="3.2" data-path="snp-and-indel-filtering.html"><a href="snp-and-indel-filtering.html#sample-filtering"><i class="fa fa-check"></i><b>3.2</b> Sample filtering</a></li>
<li class="chapter" data-level="3.3" data-path="snp-and-indel-filtering.html"><a href="snp-and-indel-filtering.html#split-snps-and-indels"><i class="fa fa-check"></i><b>3.3</b> Split SNPs and INDELs</a>
<ul>
<li class="chapter" data-level="" data-path="snp-and-indel-filtering.html"><a href="snp-and-indel-filtering.html#the-strict-method"><i class="fa fa-check"></i>The strict method</a></li>
<li class="chapter" data-level="" data-path="snp-and-indel-filtering.html"><a href="snp-and-indel-filtering.html#the-complex-method"><i class="fa fa-check"></i>The complex method</a></li>
</ul></li>
<li class="chapter" data-level="3.4" data-path="snp-and-indel-filtering.html"><a href="snp-and-indel-filtering.html#filtering-biallelic-sites"><i class="fa fa-check"></i><b>3.4</b> Filtering biallelic sites</a></li>
<li class="chapter" data-level="3.5" data-path="snp-and-indel-filtering.html"><a href="snp-and-indel-filtering.html#filter-missing-genotypes"><i class="fa fa-check"></i><b>3.5</b> Filter missing genotypes</a></li>
<li class="chapter" data-level="3.6" data-path="snp-and-indel-filtering.html"><a href="snp-and-indel-filtering.html#filtering-the-maf"><i class="fa fa-check"></i><b>3.6</b> Filtering the MAF</a></li>
<li class="chapter" data-level="3.7" data-path="snp-and-indel-filtering.html"><a href="snp-and-indel-filtering.html#annotate-variant-id"><i class="fa fa-check"></i><b>3.7</b> Annotate variant ID</a></li>
<li class="chapter" data-level="3.8" data-path="snp-and-indel-filtering.html"><a href="snp-and-indel-filtering.html#change-chromosome-names"><i class="fa fa-check"></i><b>3.8</b> Change chromosome names</a></li>
</ul></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Standardized pipelines</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="short-reads-mapping" class="section level1 hasAnchor" number="1">
<h1><span class="header-section-number">Chapter 1</span> Short reads mapping<a href="short-reads-mapping.html#short-reads-mapping" class="anchor-section" aria-label="Anchor link to header"></a></h1>
<p>Required softwares:</p>
<ul>
<li>bwa-mem2 (bwa mem can be used, but is slower)</li>
<li>samtools</li>
<li>gatk</li>
</ul>
<p>Short reads mapping have to be done with <code>bwa-mem2 mem</code> (or <code>bwa mem</code>) default parameters. The output of bwa can be pipped with <code>samtools sort</code> to output a sorted bam. The <code>-T</code> option allow to give a specific name to temporary files created for the sorting, which allows to sort multiple fasta in parallel in the same directory without any name conflicts.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb1-1"><a href="short-reads-mapping.html#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Create reference index</span></span>
<span id="cb1-2"><a href="short-reads-mapping.html#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="ex">bwa-mem2</span> index reference.fasta</span>
<span id="cb1-3"><a href="short-reads-mapping.html#cb1-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-4"><a href="short-reads-mapping.html#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Run mapping</span></span>
<span id="cb1-5"><a href="short-reads-mapping.html#cb1-5" aria-hidden="true" tabindex="-1"></a><span class="ex">bwa-mem2</span> mem <span class="at">-t</span> 20 reference.fasta IlluminaReads_1.fq/fa IlluminaReads_2.fq/fa <span class="kw">|</span> <span class="ex">samtools</span> sort <span class="at">-o</span> output.bam <span class="at">-T</span> reads.Prefix.tmp</span></code></pre></div>
<p>Once the BAM file is generated, one can add ReadGroups (required for some downstream analyses) with <code>gatk AddOrReplaceReadGroups</code>.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb2-1"><a href="short-reads-mapping.html#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Add read groups</span></span>
<span id="cb2-2"><a href="short-reads-mapping.html#cb2-2" aria-hidden="true" tabindex="-1"></a><span class="ex">gatk</span> AddOrReplaceReadGroups <span class="at">-I</span> input.bam <span class="at">-O</span> output.bam <span class="at">--RGID</span> SampleName <span class="at">--RGLB</span> SampleName <span class="at">--RGPL</span> ILLUMINA <span class="at">--RGPU</span> SampleName <span class="at">--RGSM</span> SampleName</span></code></pre></div>
<p>Finally, the bam file have to be indexed with <code>samtools index</code>.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb3-1"><a href="short-reads-mapping.html#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Index BAM</span></span>
<span id="cb3-2"><a href="short-reads-mapping.html#cb3-2" aria-hidden="true" tabindex="-1"></a><span class="ex">samtools</span> index input.bam</span></code></pre></div>

</div>
            </section>

          </div>
        </div>
      </div>
<a href="index.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="snps-and-indels-calling.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/USERNAME/REPO/edit/BRANCH/01-ShortReadsMapping.Rmd",
"text": "Edit"
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": ["_main.pdf", "_main.epub"],
"search": {
"engine": "fuse",
"options": null
},
"toc": {
"collapse": "subsection"
}
});
});
</script>

</body>

</html>
