<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 3 SNP and INDEL filtering | Standardized pipelines</title>
  <meta name="description" content="This book contains the standardized pipelines used in the HaploTeam." />
  <meta name="generator" content="bookdown 0.34 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 3 SNP and INDEL filtering | Standardized pipelines" />
  <meta property="og:type" content="book" />
  
  <meta property="og:description" content="This book contains the standardized pipelines used in the HaploTeam." />
  <meta name="github-repo" content="rstudio/bookdown-demo" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 3 SNP and INDEL filtering | Standardized pipelines" />
  
  <meta name="twitter:description" content="This book contains the standardized pipelines used in the HaploTeam." />
  

<meta name="author" content="Victor Loegler" />


<meta name="date" content="2023-07-17" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="snps-and-indels-calling.html"/>

<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.1.0/anchor-sections.css" rel="stylesheet" />
<link href="libs/anchor-sections-1.1.0/anchor-sections-hash.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.1.0/anchor-sections.js"></script>
<script src="libs/clipboard-1.7.1/clipboard.min.js"></script>
<link href="libs/primer-tooltips-1.4.0/build.css" rel="stylesheet" />
<link href="libs/klippy-0.0.0.9500/css/klippy.min.css" rel="stylesheet" />
<script src="libs/klippy-0.0.0.9500/js/klippy.min.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<style type="text/css">
  
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">HaploTeam's Standardized Pipelines</a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Introduction</a></li>
<li class="chapter" data-level="1" data-path="short-reads-mapping.html"><a href="short-reads-mapping.html"><i class="fa fa-check"></i><b>1</b> Short reads mapping</a></li>
<li class="chapter" data-level="2" data-path="snps-and-indels-calling.html"><a href="snps-and-indels-calling.html"><i class="fa fa-check"></i><b>2</b> SNPs and INDELs calling</a></li>
<li class="chapter" data-level="3" data-path="snp-and-indel-filtering.html"><a href="snp-and-indel-filtering.html"><i class="fa fa-check"></i><b>3</b> SNP and INDEL filtering</a>
<ul>
<li class="chapter" data-level="3.1" data-path="snp-and-indel-filtering.html"><a href="snp-and-indel-filtering.html#quality-filtering"><i class="fa fa-check"></i><b>3.1</b> Quality filtering</a></li>
<li class="chapter" data-level="3.2" data-path="snp-and-indel-filtering.html"><a href="snp-and-indel-filtering.html#sample-filtering"><i class="fa fa-check"></i><b>3.2</b> Sample filtering</a></li>
<li class="chapter" data-level="3.3" data-path="snp-and-indel-filtering.html"><a href="snp-and-indel-filtering.html#missing-genotype-filtering"><i class="fa fa-check"></i><b>3.3</b> Missing genotype filtering</a></li>
<li class="chapter" data-level="3.4" data-path="snp-and-indel-filtering.html"><a href="snp-and-indel-filtering.html#filter-on-excess-of-heterozygosity"><i class="fa fa-check"></i><b>3.4</b> Filter on excess of heterozygosity</a></li>
<li class="chapter" data-level="3.5" data-path="snp-and-indel-filtering.html"><a href="snp-and-indel-filtering.html#split-snps-and-indels"><i class="fa fa-check"></i><b>3.5</b> Split SNPs and INDELs</a></li>
<li class="chapter" data-level="3.6" data-path="snp-and-indel-filtering.html"><a href="snp-and-indel-filtering.html#filter-variant-position"><i class="fa fa-check"></i><b>3.6</b> Filter variant position</a></li>
<li class="chapter" data-level="3.7" data-path="snp-and-indel-filtering.html"><a href="snp-and-indel-filtering.html#filtering-biallelic-sites"><i class="fa fa-check"></i><b>3.7</b> Filtering biallelic sites</a></li>
<li class="chapter" data-level="3.8" data-path="snp-and-indel-filtering.html"><a href="snp-and-indel-filtering.html#filtering-the-maf"><i class="fa fa-check"></i><b>3.8</b> Filtering the MAF</a></li>
<li class="chapter" data-level="3.9" data-path="snp-and-indel-filtering.html"><a href="snp-and-indel-filtering.html#annotate-variant-id"><i class="fa fa-check"></i><b>3.9</b> Annotate variant ID</a></li>
<li class="chapter" data-level="3.10" data-path="snp-and-indel-filtering.html"><a href="snp-and-indel-filtering.html#change-chromosome-names"><i class="fa fa-check"></i><b>3.10</b> Change chromosome names</a></li>
</ul></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Standardized pipelines</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="snp-and-indel-filtering" class="section level1 hasAnchor" number="3">
<h1><span class="header-section-number">Chapter 3</span> SNP and INDEL filtering<a href="snp-and-indel-filtering.html#snp-and-indel-filtering" class="anchor-section" aria-label="Anchor link to header"></a></h1>
<p>This chapter contains bcftools commands to filter multi sample VCF files. These commands were tested using bcftools 1.15.1. bcftools manual: <a href="https://samtools.github.io/bcftools/bcftools.html" class="uri">https://samtools.github.io/bcftools/bcftools.html</a></p>
<p>The order in which the commands are executed must be respected. The raw VCF should first be filtered on the genotyping quality, then a subset of samples can be filtered if needed and finally several filters can be applied on the variants dataset (missing genotype filters, frequency filters, biallelic filters, etc). It is better to work with all-position VCFs, and filter variant position only in the end.</p>
<p>WARNING: Each filter must be indicated in the file name. Example: <code>full2489Samples.DP10.GQ20.SNPs.99pNonMissing.Biallelic.MAF5.vcf.gz</code></p>
<p>The different commands can be combined with <code>|</code>. To accelerate the compression of the output VCF, the command <code>bcftools view -Oz -o out.vcf.gz --threads 8</code>can be used to parallelize the compression of the file.</p>
<div id="quality-filtering" class="section level2 hasAnchor" number="3.1">
<h2><span class="header-section-number">3.1</span> Quality filtering<a href="snp-and-indel-filtering.html#quality-filtering" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>Quality filtering is the first step to process. It includes depth filtering, genotype quality or any other filters based on quality metrics (QUAL, allele balance, etc). Loci that does not pass the thresholds must be set to missing. The allele count for each loci must then be updated <code>bcftools +fill-tags</code>.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb4-1"><a href="snp-and-indel-filtering.html#cb4-1" tabindex="-1"></a><span class="co"># Depth filtering 10X</span></span>
<span id="cb4-2"><a href="snp-and-indel-filtering.html#cb4-2" tabindex="-1"></a><span class="ex">bcftools</span> +setGT in.vcf.gz <span class="at">--</span> <span class="at">-t</span> q <span class="at">-n</span> . <span class="at">-e</span> <span class="st">&#39;FMT/DP&gt;=10&#39;</span> <span class="kw">|</span> <span class="dt">\</span></span>
<span id="cb4-3"><a href="snp-and-indel-filtering.html#cb4-3" tabindex="-1"></a>  <span class="ex">bcftools</span> +fill-tags</span>
<span id="cb4-4"><a href="snp-and-indel-filtering.html#cb4-4" tabindex="-1"></a></span>
<span id="cb4-5"><a href="snp-and-indel-filtering.html#cb4-5" tabindex="-1"></a><span class="co"># Genotype quality filtering 20</span></span>
<span id="cb4-6"><a href="snp-and-indel-filtering.html#cb4-6" tabindex="-1"></a><span class="ex">bcftools</span> +setGT in.vcf.gz <span class="at">--</span> <span class="at">-t</span> q <span class="at">-n</span> . <span class="at">-e</span> <span class="st">&#39;FMT/GQ&gt;=20&#39;</span> <span class="kw">|</span> <span class="dt">\</span></span>
<span id="cb4-7"><a href="snp-and-indel-filtering.html#cb4-7" tabindex="-1"></a>  <span class="ex">bcftools</span> +fill-tags</span></code></pre></div>
</div>
<div id="sample-filtering" class="section level2 hasAnchor" number="3.2">
<h2><span class="header-section-number">3.2</span> Sample filtering<a href="snp-and-indel-filtering.html#sample-filtering" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>If a subset of sample must be extracted from the VCF, this is the second step to process. The list of sample to keep must be in txt format, 1 sample per line. Because some variants might be present in the total population but absent from the subset, multiallelic loci must be split into biallelic records before sample filtering. Then, sample are filtered and only loci with an allele count superior to 0 will be kept. After that, variants located at the same loci will be re-merged into multiallelic variants.</p>
<p>The <code>bcftools norm -m -any</code> command splits multiallelic records<br />
The <code>bcftools view --samples-file</code>command filters samples from a txt file<br />
The <code>bcftools +fill-tags</code> command updates the allele count value<br />
The <code>bcftools view --min-ac 1</code> keeps only sites with a minimum allele count of 1<br />
The <code>bcftools norm -m +any</code> command merge variant located at the same position</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb5-1"><a href="snp-and-indel-filtering.html#cb5-1" tabindex="-1"></a><span class="co"># Sample filtering</span></span>
<span id="cb5-2"><a href="snp-and-indel-filtering.html#cb5-2" tabindex="-1"></a><span class="ex">bcftools</span> norm <span class="at">-m</span> <span class="at">-any</span> in.vcf.gz <span class="kw">|</span> <span class="dt">\</span></span>
<span id="cb5-3"><a href="snp-and-indel-filtering.html#cb5-3" tabindex="-1"></a>  <span class="ex">bcftools</span> view <span class="at">--samples-file</span> samples.txt <span class="kw">|</span> <span class="dt">\</span></span>
<span id="cb5-4"><a href="snp-and-indel-filtering.html#cb5-4" tabindex="-1"></a>  <span class="ex">bcftools</span> +fill-tags <span class="kw">|</span> <span class="dt">\</span></span>
<span id="cb5-5"><a href="snp-and-indel-filtering.html#cb5-5" tabindex="-1"></a>  <span class="ex">bcftools</span> view <span class="at">--min-ac</span> 1 <span class="kw">|</span> <span class="ex">\ </span> <span class="co"># This command filter out non variant loci</span></span>
<span id="cb5-6"><a href="snp-and-indel-filtering.html#cb5-6" tabindex="-1"></a>  <span class="ex">bcftools</span> norm <span class="at">-m</span> +any</span></code></pre></div>
</div>
<div id="missing-genotype-filtering" class="section level2 hasAnchor" number="3.3">
<h2><span class="header-section-number">3.3</span> Missing genotype filtering<a href="snp-and-indel-filtering.html#missing-genotype-filtering" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>Multiple filters can be done on missing genotypes. First we can filter out samples having less than X% of informed loci. Fill-tags must be used since the number of samples changes.</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb6-1"><a href="snp-and-indel-filtering.html#cb6-1" tabindex="-1"></a><span class="co"># Get for each sample the number of loci with missing genotype</span></span>
<span id="cb6-2"><a href="snp-and-indel-filtering.html#cb6-2" tabindex="-1"></a><span class="ex">bcftools</span> stats <span class="at">-s</span> <span class="at">-</span> <span class="va">$PREFIX</span>.DP10.GQ20.vcf.gz <span class="kw">|</span> <span class="fu">grep</span> <span class="at">-E</span> ^PSC <span class="kw">|</span> <span class="fu">cut</span> <span class="at">-f3,14</span> <span class="op">&gt;</span> <span class="va">$PREFIX</span>.DP10.GQ20.imiss</span>
<span id="cb6-3"><a href="snp-and-indel-filtering.html#cb6-3" tabindex="-1"></a><span class="co"># Get the total number of sites in the VCF</span></span>
<span id="cb6-4"><a href="snp-and-indel-filtering.html#cb6-4" tabindex="-1"></a><span class="va">nSites</span><span class="op">=</span><span class="va">$(</span><span class="ex">bcftools</span> +counts <span class="va">$PREFIX</span>.DP10.GQ20.vcf.gz <span class="kw">|</span> <span class="fu">grep</span> <span class="st">&quot;Number of sites&quot;</span> <span class="kw">|</span> <span class="fu">rev</span> <span class="kw">|</span> <span class="fu">cut</span> <span class="at">-d</span> <span class="st">&quot; &quot;</span> <span class="at">-f</span> 1 <span class="kw">|</span> <span class="fu">rev</span><span class="va">)</span></span>
<span id="cb6-5"><a href="snp-and-indel-filtering.html#cb6-5" tabindex="-1"></a><span class="co"># List samples with less than 20% missing genotypes</span></span>
<span id="cb6-6"><a href="snp-and-indel-filtering.html#cb6-6" tabindex="-1"></a><span class="fu">awk</span> <span class="at">-v</span> nSites=<span class="va">$nSites</span> <span class="st">&#39;{if ($2 / nSites &lt;= 0.2) print $1}&#39;</span> <span class="va">$PREFIX</span>.DP10.GQ20.imiss <span class="op">&gt;</span> Samples.Mind20</span>
<span id="cb6-7"><a href="snp-and-indel-filtering.html#cb6-7" tabindex="-1"></a><span class="co"># Filter samples</span></span>
<span id="cb6-8"><a href="snp-and-indel-filtering.html#cb6-8" tabindex="-1"></a><span class="ex">bcftools</span> view <span class="at">--samples-file</span> Samples.Mind20 <span class="va">$PREFIX</span>.DP10.GQ20.vcf.gz <span class="kw">|</span> <span class="dt">\</span></span>
<span id="cb6-9"><a href="snp-and-indel-filtering.html#cb6-9" tabindex="-1"></a>    <span class="ex">bcftools</span> +fill-tags</span></code></pre></div>
<p>Then we can filter loci that are not informed for enough samples (ex: filter out loci that are informed for less than 99% of the samples).</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb7-1"><a href="snp-and-indel-filtering.html#cb7-1" tabindex="-1"></a><span class="co"># Keep only sites with more than 99% of genotyped samples</span></span>
<span id="cb7-2"><a href="snp-and-indel-filtering.html#cb7-2" tabindex="-1"></a><span class="ex">bcftools</span> view <span class="at">-i</span> <span class="st">&#39;F_MISSING&lt;0.01&#39;</span></span></code></pre></div>
</div>
<div id="filter-on-excess-of-heterozygosity" class="section level2 hasAnchor" number="3.4">
<h2><span class="header-section-number">3.4</span> Filter on excess of heterozygosity<a href="snp-and-indel-filtering.html#filter-on-excess-of-heterozygosity" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>Regions duplicated in some samples but in a single copy in the reference genome will often result in excess of heterozygosity. These regions can therefore be filtered with the ExcHet field, which tests for excess of heterozygosity and is calculated with fill-tags. This filter is usefull to avoid artificial increase in diversity caused by duplicated regions.</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb8-1"><a href="snp-and-indel-filtering.html#cb8-1" tabindex="-1"></a><span class="ex">bcftools</span> +fill-tags <span class="kw">|</span> <span class="ex">bcftools</span> view <span class="at">-e</span> <span class="st">&#39;ExcHet &lt; 0.99&#39;</span></span></code></pre></div>
</div>
<div id="split-snps-and-indels" class="section level2 hasAnchor" number="3.5">
<h2><span class="header-section-number">3.5</span> Split SNPs and INDELs<a href="snp-and-indel-filtering.html#split-snps-and-indels" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>In raw VCFs, SNPs and INDELs can be hard to discriminate because they can be located at the same loci (called mixed loci). Some raw VCFs also contains MNVs (= Multiple Nucleotide Variants), that can be broken down in multiple consecutive SNPs. This makes the dicrimination between SNPs and INDELs very complex. Since those mixed loci are hard to analyse, we filter them out. The raw VCF will be split in 2 VCF: one containing only strict SNPs (no INDEL present at these loci in the population), one containing only strict INDELs (no SNP present at these loci in the population). Non variant positions, refered as <code>REF</code> type will be kept.</p>
<p>For SNPs, all sites with <code>*</code> as genotype (= position is overlapping a deletion) or not <code>SNP</code> or <code>REF</code> as type will be excluded with the command <code>'ALT="*" || (type!="snp" &amp;&amp; type!="ref")'</code>. Same for INDELs, except <code>*</code>gontypes are kept: <code>'(type!="indel" &amp;&amp; type!="ref")'</code>.</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb9-1"><a href="snp-and-indel-filtering.html#cb9-1" tabindex="-1"></a><span class="co"># Keep only SNPs</span></span>
<span id="cb9-2"><a href="snp-and-indel-filtering.html#cb9-2" tabindex="-1"></a><span class="ex">bcftools</span> view <span class="at">-e</span> <span class="st">&#39;ALT=&quot;*&quot; || (type!=&quot;snp&quot; &amp;&amp; type!=&quot;ref&quot;)&#39;</span></span>
<span id="cb9-3"><a href="snp-and-indel-filtering.html#cb9-3" tabindex="-1"></a></span>
<span id="cb9-4"><a href="snp-and-indel-filtering.html#cb9-4" tabindex="-1"></a><span class="co"># Keep only INDELs</span></span>
<span id="cb9-5"><a href="snp-and-indel-filtering.html#cb9-5" tabindex="-1"></a><span class="ex">bcftools</span> view <span class="at">-e</span> <span class="st">&#39;(type!=&quot;indel&quot; &amp;&amp; type!=&quot;ref&quot;)&#39;</span></span></code></pre></div>
</div>
<div id="filter-variant-position" class="section level2 hasAnchor" number="3.6">
<h2><span class="header-section-number">3.6</span> Filter variant position<a href="snp-and-indel-filtering.html#filter-variant-position" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>Non variant position can be filtered out based on allele count (AC - must be up to date using fill-tags).</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb10-1"><a href="snp-and-indel-filtering.html#cb10-1" tabindex="-1"></a><span class="co"># Keep only variant positions</span></span>
<span id="cb10-2"><a href="snp-and-indel-filtering.html#cb10-2" tabindex="-1"></a><span class="ex">bcftools</span> view <span class="at">--min-ac</span> 1</span></code></pre></div>
</div>
<div id="filtering-biallelic-sites" class="section level2 hasAnchor" number="3.7">
<h2><span class="header-section-number">3.7</span> Filtering biallelic sites<a href="snp-and-indel-filtering.html#filtering-biallelic-sites" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>Biallelic loci can be kept with the <code>bcftools view</code> command, keeping loci with minimum 2 alleles (<code>-m2</code>) and maximum 2 alleles (<code>-M2</code>).</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb11-1"><a href="snp-and-indel-filtering.html#cb11-1" tabindex="-1"></a><span class="co"># Filter biallelic (minimum and maximum 2 alleles per loci)</span></span>
<span id="cb11-2"><a href="snp-and-indel-filtering.html#cb11-2" tabindex="-1"></a><span class="ex">bcftools</span> view <span class="at">-m2</span> <span class="at">-M2</span> in.vcf.gz</span></code></pre></div>
</div>
<div id="filtering-the-maf" class="section level2 hasAnchor" number="3.8">
<h2><span class="header-section-number">3.8</span> Filtering the MAF<a href="snp-and-indel-filtering.html#filtering-the-maf" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<div class="sourceCode" id="cb12"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb12-1"><a href="snp-and-indel-filtering.html#cb12-1" tabindex="-1"></a><span class="co"># Sites with a MAF lower are removed</span></span>
<span id="cb12-2"><a href="snp-and-indel-filtering.html#cb12-2" tabindex="-1"></a><span class="ex">bcftools</span> view <span class="at">--min-af</span> 0.05:minor</span></code></pre></div>
</div>
<div id="annotate-variant-id" class="section level2 hasAnchor" number="3.9">
<h2><span class="header-section-number">3.9</span> Annotate variant ID<a href="snp-and-indel-filtering.html#annotate-variant-id" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>Variants ID can be annotated at any step of the filtering process, but it’s generally the lkast step of the pipeline. ID annotation must be done to be able to discriminate variants.</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb13-1"><a href="snp-and-indel-filtering.html#cb13-1" tabindex="-1"></a><span class="co"># Set variant ID to rs_Chr_Pos_etc...</span></span>
<span id="cb13-2"><a href="snp-and-indel-filtering.html#cb13-2" tabindex="-1"></a><span class="ex">bcftools</span> annotate <span class="at">--set-id</span> <span class="st">&#39;rs_%CHROM\_%POS\_%REF\_%FIRST_ALT&#39;</span> </span></code></pre></div>
</div>
<div id="change-chromosome-names" class="section level2 hasAnchor" number="3.10">
<h2><span class="header-section-number">3.10</span> Change chromosome names<a href="snp-and-indel-filtering.html#change-chromosome-names" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>Some software such as PLINK requires integers as chromosome names. Chromosome names can be changed with <code>bcftools annotate --rename-chrs</code></p>
<div class="sourceCode" id="cb14"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb14-1"><a href="snp-and-indel-filtering.html#cb14-1" tabindex="-1"></a><span class="co"># Create file that will contain name substitutions</span></span>
<span id="cb14-2"><a href="snp-and-indel-filtering.html#cb14-2" tabindex="-1"></a><span class="kw">(</span></span>
<span id="cb14-3"><a href="snp-and-indel-filtering.html#cb14-3" tabindex="-1"></a>  <span class="bu">echo</span> <span class="st">&quot;chromosome1 1&quot;</span></span>
<span id="cb14-4"><a href="snp-and-indel-filtering.html#cb14-4" tabindex="-1"></a>  <span class="bu">echo</span> <span class="st">&quot;chromosome2 2&quot;</span></span>
<span id="cb14-5"><a href="snp-and-indel-filtering.html#cb14-5" tabindex="-1"></a>  <span class="bu">echo</span> <span class="st">&quot;chromosome3 3&quot;</span></span>
<span id="cb14-6"><a href="snp-and-indel-filtering.html#cb14-6" tabindex="-1"></a>  <span class="bu">echo</span> <span class="st">&quot;chromosome4 4&quot;</span></span>
<span id="cb14-7"><a href="snp-and-indel-filtering.html#cb14-7" tabindex="-1"></a>  <span class="bu">echo</span> <span class="st">&quot;chromosome5 5&quot;</span></span>
<span id="cb14-8"><a href="snp-and-indel-filtering.html#cb14-8" tabindex="-1"></a>  <span class="bu">echo</span> <span class="st">&quot;chromosome6 6&quot;</span></span>
<span id="cb14-9"><a href="snp-and-indel-filtering.html#cb14-9" tabindex="-1"></a>  <span class="bu">echo</span> <span class="st">&quot;chromosome7 7&quot;</span></span>
<span id="cb14-10"><a href="snp-and-indel-filtering.html#cb14-10" tabindex="-1"></a>  <span class="bu">echo</span> <span class="st">&quot;chromosome8 8&quot;</span></span>
<span id="cb14-11"><a href="snp-and-indel-filtering.html#cb14-11" tabindex="-1"></a>  <span class="bu">echo</span> <span class="st">&quot;chromosome9 9&quot;</span></span>
<span id="cb14-12"><a href="snp-and-indel-filtering.html#cb14-12" tabindex="-1"></a>  <span class="bu">echo</span> <span class="st">&quot;chromosome10 10&quot;</span></span>
<span id="cb14-13"><a href="snp-and-indel-filtering.html#cb14-13" tabindex="-1"></a>  <span class="bu">echo</span> <span class="st">&quot;chromosome11 11&quot;</span></span>
<span id="cb14-14"><a href="snp-and-indel-filtering.html#cb14-14" tabindex="-1"></a>  <span class="bu">echo</span> <span class="st">&quot;chromosome12 12&quot;</span></span>
<span id="cb14-15"><a href="snp-and-indel-filtering.html#cb14-15" tabindex="-1"></a>  <span class="bu">echo</span> <span class="st">&quot;chromosome13 13&quot;</span></span>
<span id="cb14-16"><a href="snp-and-indel-filtering.html#cb14-16" tabindex="-1"></a>  <span class="bu">echo</span> <span class="st">&quot;chromosome14 14&quot;</span></span>
<span id="cb14-17"><a href="snp-and-indel-filtering.html#cb14-17" tabindex="-1"></a>  <span class="bu">echo</span> <span class="st">&quot;chromosome15 15&quot;</span></span>
<span id="cb14-18"><a href="snp-and-indel-filtering.html#cb14-18" tabindex="-1"></a>  <span class="bu">echo</span> <span class="st">&quot;chromosome16 16&quot;</span><span class="kw">)</span> <span class="op">&gt;</span> ChrNames.txt</span>
<span id="cb14-19"><a href="snp-and-indel-filtering.html#cb14-19" tabindex="-1"></a><span class="co"># Change chr names with bcftools</span></span>
<span id="cb14-20"><a href="snp-and-indel-filtering.html#cb14-20" tabindex="-1"></a><span class="ex">bcftools</span> annotate <span class="at">--rename-chrs</span> ChrNames.txt in.vcf.gz</span></code></pre></div>

</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="snps-and-indels-calling.html" class="navigation navigation-prev navigation-unique" aria-label="Previous page"><i class="fa fa-angle-left"></i></a>

    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/USERNAME/REPO/edit/BRANCH/03-SNPfiltering.Rmd",
"text": "Edit"
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": ["_main.pdf", "_main.epub"],
"search": {
"engine": "fuse",
"options": null
},
"toc": {
"collapse": "subsection"
}
});
});
</script>

</body>

</html>
